lib/symboldatabase.cpp

```diff
@@ -2272,20 +2272,27 @@ void Variable::setValueType(const ValueType &valueType)
         setFlag(fIsSmartPointer, true);
 }
 
-const Type *Variable::smartPointerType() const
+const Type* Variable::smartPointerType() const
 {
     if (!isSmartPointer())
         return nullptr;
 
     if (mValueType->smartPointerType)
         return mValueType->smartPointerType;
 
-    // TODO: Cache result
-    const Token *ptrType = typeStartToken();
-    while (Token::Match(ptrType, "%name%|::"))
-        ptrType = ptrType->next();
-    if (Token::Match(ptrType, "< %name% >"))
-        return ptrType->scope()->findType(ptrType->next()->str());
+    // TODO: Cache result, handle more complex type expression
+    const Token* typeTok = typeStartToken();
+    while (Token::Match(typeTok, "%name%|::"))
+        typeTok = typeTok->next();
+    if (Token::Match(typeTok, "< %name% >")) {
+        const Scope* scope = typeTok->scope();
+        const Type* ptrType{};
+        while (scope && !ptrType) {
```

Comment:
We're in a `Variable` member function here, so I don't think that's accessible.

Reply:
hmm.. the reason as far as I see that is put in SymbolDatabase is because it uses SymbolDatabase::mTokenizer to determine if the code is C code or not.

maybe we could have a function somewhere that takes the language as parameter. I would assume we know it's C++ code in this `Variable::smartPointerType` method.

Would it make sense to create a static `const Type* SymbolDatabase::findType(const Token *startTok, const Scope *startScope, bool lookOutside, bool isC)` method? That is called from the non-static SymbolDatabase::findType methods.

Reply:
There's also usage of the `scopeList` member, but I tried it anyway. Turns out that `SymbolDatabase` doesn't really iterate over scopes like the other code does, so the new test case fails.

Reply:
ok thanks that you tried it out at least!